pkg/resourcemanager/predicate/class_filter.go

```diff
@@ -76,48 +79,39 @@ func (f *ClassFilter) Responsible(o runtime.Object) bool {
 	return c == f.resourceClass || (c == "" && f.resourceClass == resourcemanagerv1alpha1.DefaultResourceClass)
 }
 
-// Active checks whether a dedicated object must be handled by the actual controller
-// instance. This is split into two conditions. An object must be handled
-// if it has already been handled, indicated by the actual finalizer, or
-// if the actual controller is responsible for the object.
-func (f *ClassFilter) Active(o runtime.Object) (action bool, responsible bool) {
-	busy := false
-	responsible = f.Responsible(o)
-	r := o.(*resourcesv1alpha1.ManagedResource)
+// NoLongerHandles checks if an MR has changed its class and should have its resources cleaned by the given controller instance.
+func (f *ClassFilter) NoLongerHandles(mr *resourcesv1alpha1.ManagedResource) bool {
+	return controllerutil.ContainsFinalizer(mr, f.objectFinalizer) && !f.Responsible(mr)
+}
 
-	for _, finalizer := range r.GetFinalizers() {
+// WaitForCleanup checks if a MR has changed its class and a given controller instance should wait for its resources to be cleaned.
+func (f *ClassFilter) WaitForCleanup(mr *resourcesv1alpha1.ManagedResource) bool {
```

Comment:
```suggestion
func (f *ClassFilter) IsWaitForCleanupRequired(mr *resourcesv1alpha1.ManagedResource) bool {
```
`WaitForCleanup` implies that the function will do the actual wait.